worker/work.go

```diff
@@ -0,0 +1,203 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/elastic/hey-apm/conv"
+	"github.com/elastic/hey-apm/numbers"
+	"github.com/elastic/hey-apm/out"
+
+	"github.com/heptio/workgroup"
+
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/stacktrace"
+)
+
+type Worker struct {
+	*out.ApmLogger
+	*apm.Tracer
+	RunTimeout   time.Duration
+	FlushTimeout time.Duration
+
+	// not to be modified concurrently
+	workgroup.Group
+}
+
+type metric struct {
+	Name, Value string
+}
+
+type Report struct {
+	Stats   []metric
+	Start   time.Time
+	End     time.Time
+	Flushed time.Time
+}
+
+func (r *Report) add(metricName string, value interface{}) {
+	r.Stats = append(r.Stats, metric{metricName, conv.StringOf(value)})
+}
+
+func (w *Worker) Work() (Report, error) {
+	if w.RunTimeout > 0 {
+		w.Add(func(done <-chan struct{}) error {
+			select {
+			case <-done:
+				return nil
+			case <-time.After(w.RunTimeout):
+				return nil // time expired
+			}
+		})
+	}
+
+	report := Report{Stats: make([]metric, 0)}
+	report.Start = time.Now()
+	err := w.Run()
+	report.End = time.Now()
+	w.flush()
+	report.Flushed = time.Now()
+
+	rs := w.Stats()
+	report.add("transactions sent", rs.TransactionsSent)
+	report.add("transactions dropped", rs.TransactionsDropped)
+	if rs.TransactionsSent+rs.TransactionsDropped > 0 {
+		report.add("transactions sent Perct.", numbers.Perct(rs.TransactionsSent, rs.TransactionsDropped))
+		report.add("spans sent", rs.SpansSent)
+		report.add("spans dropped", rs.SpansDropped)
+		report.add("spans sent Perct.", numbers.Perct(rs.SpansSent, rs.SpansDropped))
+		if rs.TransactionsSent > 0 {
+			report.add("spans sent per transaction", numbers.Div(rs.SpansSent, rs.TransactionsSent))
+		}
+	}
+	report.add("errors sent", rs.ErrorsSent)
+	report.add("errors dropped", rs.ErrorsDropped)
+	if rs.ErrorsSent+rs.ErrorsDropped > 0 {
+		report.add("errors sent Perct.", numbers.Perct(rs.ErrorsSent, rs.ErrorsDropped))
+	}
+	report.add("failed", rs.Errors.SendStream)
+	eventsSent := float64(rs.ErrorsSent + rs.SpansSent + rs.TransactionsSent)
+	report.add("events sent per second", eventsSent/report.Flushed.Sub(report.Start).Seconds())
```

Comment:
worth it to include total events sent too?  Maybe combine with the per second line like `events sent .... 123 (12.5/s)`

Reply:
done in https://github.com/elastic/hey-apm/pull/102/commits/714ac272827a7cc1bfb467dd8f3f9b4dcdb2d70f

(i never know when or not to split up PRs...)